### PR TITLE
Bridges - changes for Bridges V2 - relay client part

### PR DESCRIPTION
**Pull request by bkontur**
Contains mainly changes/nits/refactors related to the relayer code (`client-substrate` and `lib-substrate-relay`) migrated from the Bridges V2 [branch](https://github.com/paritytech/polkadot-sdk/pull/4427).

Relates to: https://github.com/paritytech/parity-bridges-common/issues/2976
Companion: https://github.com/paritytech/parity-bridges-common/pull/2988

## TODO
- [x] fix comments

## Questions
- [x] Do we need more testing for client V2 stuff? If so, how/what is the ultimate test? @svyatonik 
  - [x] check [comment](https://github.com/paritytech/polkadot-sdk/pull/4494#issuecomment-2117181144) for more testing

### CODE REVIEW

**Comment by svyatonik** [comment]
@svyatonik In traitified v2 client we are using shared subscriptions, so https://github.com/paritytech/polkadot-sdk/pull/4481 shouldn't be a problem when we talk about long-lived subscriptions (justification). But we need to test it for transaction subscriptions - right now I've rebased completely without #4481

**Review by svyatonik — COMMENTED** [review]
FWIW: all code changes were already reviewed before in a https://github.com/paritytech/parity-bridges-common/ repository. I did a shallow check and porting looks good and should work with all existing bridges. However I also want to perform that check: https://github.com/paritytech/polkadot-sdk/pull/4494#issuecomment-2117181144. Will do that tomorrow and approve if all is ok or push some commits to fix it if changes are required

**Review by svyatonik — APPROVED** [review]
Checked the possible subscription leak issue - all is good

File: bridges/relays/client-substrate/src/client/caching.rs [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+	}
+}
+
+impl<C: Chain, B: Client<C>> std::fmt::Debug for CachingClient<C, B> {
+	fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+		fmt.write_fmt(format_args!("CachingClient<{:?}>", self.backend))
+	}
+}
+
+// TODO (https://github.com/paritytech/parity-bridges-common/issues/2133): this must be implemented for T: Client<C>
```
Comment (acatangiu) [inline bridges/relays/client-substrate/src/client/caching.rs:257]: Is this still relevant?

File: bridges/relays/client-substrate/src/client/mod.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+use num_traits::Saturating;
+use rpc::RpcClient;
+use sp_version::RuntimeVersion;
+
+pub mod caching;
+pub mod rpc;
+
+// don't want to move all the trait code (200+ lines) here and there's no better name
+#[allow(clippy::module_inception)]
+mod client;
```
Comment (acatangiu) [inline bridges/relays/client-substrate/src/client/mod.rs:31]:
```suggestion
mod traits;
```

File: bridges/relays/client-substrate/src/transaction_tracker.rs [hunk 3]
```diff
@@ -76,6 +78,21 @@ impl<C: Chain, E: Environment<C>> TransactionTracker<C, E> {
 		Self { environment, stall_timeout, transaction_hash, subscription }
 	}
 
+	// TODO (https://github.com/paritytech/parity-bridges-common/issues/2133): remove me after
+	// test client is implemented
```
Comment (acatangiu) [inline bridges/relays/client-substrate/src/transaction_tracker.rs:82]: if we still need this, pls let's create a more descriptive issue for it. It's not clear what the preconditions for this TODO are.

**Review by acatangiu — APPROVED** [review]
Did shallow review only since this has already been reviewed on development branch

File: bridges/relays/client-substrate/src/client/mod.rs (hunk 2 above)
Comment (bkontur) [reply bridges/relays/client-substrate/src/client/mod.rs]: fixed here: https://github.com/paritytech/polkadot-sdk/pull/4494/commits/c0eeff62f3862df59b456b03b80803bdac48db07

File: bridges/relays/client-substrate/src/transaction_tracker.rs (hunk 3 above)
Comment (bkontur) [reply bridges/relays/client-substrate/src/transaction_tracker.rs:82]: I summarized all TODOs here https://github.com/paritytech/parity-bridges-common/issues/3009 (I also pinged Slava)

**Review by serban300 — APPROVED** [review]
Also took just a high-level look since this has already been reviewed before. Looks good !

**Pull request merged**